airflow/executors/inprocess_executor.py

```diff
@@ -0,0 +1,152 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+This module contains InProcessExecutor that is a single
+process executor meaning it does not use multiprocessing.
+"""
+
+import threading
+from typing import Any, Dict, List, Optional
+
+from airflow import conf
+from airflow.executors.base_executor import BaseExecutor
+from airflow.models.taskinstance import TaskInstance, TaskInstanceKeyType
+from airflow.utils.state import State
+
+
+class InProcessExecutor(BaseExecutor):
+    """
+    This executor is meant for debugging purposes. It can be used with
+    sqlite since sqlite do not support multiple connections.
+
+    It executes one task instance at time. Additionally to support working
+    with sensors, sensor's ``mode`` will be automatically set to "reschedule".
+    """
+
+    _terminated = threading.Event()
+
+    def __init__(self):
+        super().__init__()
+        self.tasks_to_run: List[TaskInstance] = []
+        # Place where we keep information for task instance raw run
+        self.tasks_params: Dict[TaskInstanceKeyType, Dict[str, Any]] = {}
+        self.fail_fast = conf.getboolean("debug", "fail_fast")
+
+    def execute_async(self, *args, **kwargs) -> None:
+        """
+        The method is replaced by custom trigger_task implementation.
+        """
+
+    def sync(self) -> None:
+        task_succeeded = True
+        while self.tasks_to_run:
+            ti = self.tasks_to_run.pop(0)
+            if self.fail_fast and not task_succeeded:
+                self.log.info("Setting %s to %s", ti.key, State.UPSTREAM_FAILED)
+                ti.set_state(State.UPSTREAM_FAILED)
+                self.change_state(ti.key, State.UPSTREAM_FAILED)
+                continue
+
+            if self._terminated.is_set():
+                self.log.info(
+                    "Executor is terminated! Stopping %s to %s", ti.key, State.FAILED
+                )
+                ti.set_state(State.FAILED)
+                self.change_state(ti.key, State.FAILED)
+                continue
+
+            task_succeeded = self._run_task(ti)
+
+    def _run_task(self, ti: TaskInstance) -> bool:
+        self.log.debug("Executing task: %s", ti)
+        key = ti.key
+        try:
+            params = self.tasks_params.get(ti.key, {})
```

Comment:
Doesn't look like we ever remove entries form `task_params` - we probably should.